src/controllers/orderController.js

```diff
@@ -5,58 +5,103 @@ const { BadRequestError } = require('../errors');
+    const user = req.user;
 
-  // Perform necessary checks and validations
-  if (!orderItems?.length) {
-    throw new BadRequestError(
-      'Unable to create order because the order items are missing.'
-    );
-  }
 
-  // Create an Order document in the database
-  const order = await Order.create({
-    user: req.user.userId,
-    orderItems,
-    subtotal,
-    tax,
-    total,
-  });
 
-  // Create a payment intent with Stripe
-  const paymentIntent = await stripe.paymentIntents.create({
-    amount: total * 100, // Stripe requires amount in cents
-    currency: 'usd',
-    description: `Order #${order._id}`,
-    metadata: {
-      userId: req.user.userId,
-      orderId: order._id, // key-values
-      totalQuantity: orderItems.reduce(
-        (total, item) => total + item.quantity,
-        0
-      ), // the total number of items in the order
-      totalAlbums: orderItems.length, // the total number of albums purchased in order
-    },
-  });
 
-  // Save the paymentIntent to the Order model
-  order.paymentIntentId = paymentIntent.id;
-  await order.save();
 
-  // Create PurchasedAlbum entries for each album in orderItems
-  for (const orderItem of orderItems) {
-    await PurchasedAlbum.create({
-      album: orderItem.album, // album ID from orderItem
-      user: req.user.userId, // user ID from the request
+// Populate the orderItems with full album data
+const populatedOrderItems = await Promise.all(
+  orderItems.map(async (orderItem) => {
+    return await orderItem.populate('album').execPopulate();
+  })
+);
+  // Create an array to store the order items with full album data
+  const orderItemsWithFullAlbum = populatedOrderItems.map((orderItem) => ({
+    album: orderItem.album.toObject(), // Use toObject() to convert the full album to a plain JavaScript object
+    quantity: orderItem.quantity,
+  }));
+
+
+
+
+    const order = new Order({
+      user: req.user.userId,
+      orderItems: orderItemsWithFullAlbum,// Use the array with full album data
+      subtotal,
+      tax,
+      total,
     });
-  }
 
-  // Send the payment intent client secret and order information to the client
-  res
-    .status(StatusCodes.CREATED)
-    .json({ clientSecret: paymentIntent.client_secret, order });
+     await order.save({ session });
+
+   
+    const paymentIntent = await stripe.paymentIntents.create({
+      amount: total * 100,
+      currency: 'usd',
+      description: `Order #${order._id}`,
+      metadata: {
+        userId: req.user.userId,
+        orderId: order._id,
+        totalQuantity: orderItems.reduce(
+          (total, item) => total + item.quantity,
+          0
+        ),
+        totalAlbums: orderItems.length,
+      },
+    });
+
+    order.paymentIntentId = paymentIntent.id;
+
+    for (const orderItem of orderItems) {
+      const purchasedAlbum = new PurchasedAlbum({
+        album: orderItem.album,
+        user: req.user.userId,
+      });
+
+      await purchasedAlbum.save({ session });
+    }
```

Comment:
Let's move this block above the stripe block; that way we know we've done as much of the database saving as possible before trying to save to stripe (since the stripe part cannot be easily reversed here if the transaction fails; while all the db queries _will_ be reversed if the transaction fails)

Reply:
@akosasante just to be sure- so if this part fails 
![image](https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/103755090/67106572-f24d-4125-981f-dee6b9e44453)
setInterval inOrder.js will trigger that this order will be deleted both from Order collection and PurchasedAlbum collection?
![image](https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/103755090/82706410-616c-42cf-80bd-58aef282dab8)

Reply:
If the `stripe.paymentIntents.create` part fails from your first screenshot, that will raise an error. 

Since we are in a `try/catch` block, that means Node will proceed to the `catch` block on line 69-76. 

Then, on line 71 we tell mongo to `abortTransaction` which means it will "rollback" (that is, undo) all database changes that occurred in this transaction/session (so everything after line 11 where we started the transaction, this includes undoing the order creation and the purchased album creation). 

The `setInterval` function in `Order.js` does not come into play here at all. That function is happening independently just checking periodically for any orders that have been in `'pending'` state for too long. It updates those old pending orders to `'cancelled'`. Then separately, because there is a TTL (time-to-live) index on the `orderStatus` column, any orders that are in `'cancelled'` state for too long (currently set to 7200 seconds/2 hours) are deleted completely from the database.

Reply:
@akosasante THANK YOU!!!!